lib/manifest.ex

```diff
@@ -1,18 +1,195 @@
+          rollbacks: [Step.rollback()],
+          halt?: boolean(),
+          errored: Step.operation(),
+          reason: any()
+        }
+
+  @doc """
+  Initializes a new `t:Manifest.t()`.
+
+  The manifest is not customizable, so no options can be given to this function.
+  """
+  @spec new :: t()
+  def new, do: %__MODULE__{}
+
+  @doc """
+  Adds a `Manifest.Step` to the `:steps` key of the manifest. 
+
+  True to Elixir/Erlang practices, it is prepended to the list. The list is 
+  reversed the work is actually performed. 
+
+  See `Manifest.Step` for more information on what a step is.
+  """
+
+  @spec add_step(
+          t(),
+          atom(),
+          __MODULE__.Step.work(),
+          __MODULE__.Step.rollback(),
+          __MODULE__.Step.parser()
+        ) :: t()
+  def add_step(
+        manifest,
+        operation,
+        work,
+        rollback \\ &Step.safe_default_rollback/1,
+        parser \\ &Step.default_parser/1
+      )
+
+  def add_step(_manifest, operation, _work, _rollback, _parser) when not is_atom(operation),
+    do: raise(NotAnAtomError, operation)
+
+  def add_step(_manifest, _operation, work, _rollback, _parser) when not is_function(work),
+    do: raise(NotAFunctionError, key: :work, value: work)
+
+  def add_step(_manifest, _operation, _work, rollback, _parser) when not is_function(rollback),
+    do: raise(NotAFunctionError, key: :rollback, value: rollback)
+
+  def add_step(_manifest, _operation, _work, _rollback, parser) when not is_function(parser),
+    do: raise(NotAFunctionError, key: :parser, value: parser)
+
+  def add_step(manifest, operation, work, rollback, parser) do
+    step = step(operation: operation, work: work, parser: parser, rollback: rollback)
+    Map.update(manifest, :steps, [step], &[step | &1])
+  end
+
+  @doc """
+  Performs the steps in the order given to the manifest.
+
+  (Each new call of `add_step/1` prepends a `t:Manifest.Step.t/0` to the list,
+  then the list is reversed before performing.) If a step's
+  `t:Manifest.Step.work/0` succeeds without returning :no_rollback in the tuple,
+  the `t:Manifest.Step.parser/0` finds the identifier, and a
+  `t:Manifest.Step.rollback/0` is defined the rollback will be added to the
+  `:rollbacks` stack. If a step fails, the `:halt?` will activate and no further
+  work will be performed. The `:errored` key will be set to the operation that
+  triggered the failure. Whether or not the step fails, a key-value pair will
+  be added to the `:previous` field. The `t:Manifest.Step.operation/0` will be
+  the key of that key-value pair.
+
+  See `digest/1` as it provides an easier way of extracting pertinent 
+  information on what happened during this function. 
+  """
+  @spec perform(t()) :: t()
+  def perform(%__MODULE__{steps: steps} = manifest), do: perform(Enum.reverse(steps), manifest)
+
+  @doc """
+  Reports on the results of `perform/1`.
+
+  Returns an `:ok` tuple will the value of the `:previous` key which contains 
```

Comment:
think you mean: "with" the value of the..

Reply:
Yup thanks.